eng/pipelines/templates/jobs/ci.yml

```diff
@@ -91,25 +90,27 @@ jobs:
         inputs:
           versionSpec: '3.6'
 
-      - ${{ parameters.PreBuildSteps }}
-
       # This step needs to run before we update to using dev versions otherwise dependency
       # check in the script will not fully work because the versions will mismatch
-      - task: PythonScript@0
-        displayName: 'Generate directories variable for sparse checkout'
+      - task: PowerShell@2
+        displayName: Generate directories variable for sparse checkout
         inputs:
-          scriptPath: 'eng/scripts/generate_from_source_pom.py'
-          arguments: '--set-pipeline-variable CheckoutDirectories --project-list $(ProjectList)'
-          workingDirectory: '$(System.DefaultWorkingDirectory)'
+          pwsh: true
+          filePath: $(Build.SourcesDirectory)/eng/scripts/Generate-ServiceDirectories-From-Project-List.ps1
+          arguments: >
+            -SourcesDirectory $(Build.SourcesDirectory)
+            -ProjectList $(ProjectList)
 
       # Skip sparse checkout for the `azure-sdk-for-<lang>-pr` private mirrored repositories
       # as we require the github service connection to be loaded.
       - ${{ if not(contains(variables['Build.DefinitionName'], 'java-pr')) }}:
         - template: /eng/common/pipelines/templates/steps/sparse-checkout.yml
           parameters:
-            Paths: $(CheckoutDirectories)
+            Paths: $(SparseCheckoutDirectories)
             SkipDefaultCheckout: true
 
+      - ${{ parameters.PreBuildSteps }}
```

Comment:
Did you intent to move the PreBuildSteps after the sparse checkout?

Reply:
This was intentional, otherwise there's a chicken or egg problem with the pre build scripts not being checked out with the initial sparse checkout pass.